lib/auth/gds_sso.rb

```diff
@@ -0,0 +1,39 @@
+module Auth
+  class GdsSso
+    def self.locate(token_string)
+      Services.cache.fetch(['api-user-cache', token_string], expires_in: 5.minutes) do
+        new(token_string).locate
+      end
+    end
+
+    def initialize(token_string)
+      @token_string = token_string
+    end
+
+    def locate
+      access_token = OAuth2::AccessToken.new(oauth_client, token_string)
+      body = access_token.get("/user.json?client_id=#{CGI.escape(oauth_id)}").body
+      JSON.parse(body)["user"]
+    rescue OAuth2::Error
+      nil
```

Comment:
Could this result in `nil` being cached as the `api-user-cache`?

> Nil values can be cached ([source](https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html)).

Reply:
Yes, but only for five minutes. I think the point being that if the access token doesn't exist then we don't want to be constantly querying Signon over that period.

Reply:
In that case, would it be worth sending this authentication error to monitoring/elsewhere? It might end up being permanently cached as nil if the token becomes invalid, but would this be noticeable?

There may be cases where the writes fail due to a broken token, but once the token is fixed there could be a 5 minute delay.

Reply:
> There may be cases where the writes fail due to a broken token, but once the token is fixed there could be a 5 minute delay.

Yup that's correct, but I don't think that's too big a problem and is consistent with how GDS-SSO works: https://github.com/alphagov/gds-sso/blob/master/lib/gds-sso/bearer_token.rb#L8-L12

It'd be consistent that the monitoring of the usage is handled by the consumer application rather than here.